homeassistant/components/wiffi/__init__.py

```diff
@@ -0,0 +1,233 @@
+                self._known_devices[device.mac_address].add(metric.id)
+                async_dispatcher_send(
+                    self._hass, CREATE_ENTITY_SIGNAL, self, device, metric
+                )
+            else:
+                async_dispatcher_send(
+                    self._hass,
+                    UPDATE_ENTITY_SIGNAL + generateUniqueId(device, metric),
+                    device,
+                    metric,
+                )
+
+    @property
+    def server(self):
+        """Return TCP server instance for start + close."""
+        return self._server
+
+    @property
+    def async_add_entities(self):
+        """Return dict with add_entities functions for every platform."""
+        return self._async_add_entities
+
+    @callback
+    def _periodic_tick(self, now=None):
+        """Check if any entity has timed out because it has not been updated."""
+        async_dispatcher_send(self._hass, CHECK_ENTITIES_SIGNAL)
+
+
+class WiffiEntity(Entity):
+    """Common functionality for all wiffi entities."""
+
+    def __init__(self, device, metric):
+        """Initialize the base elements of a wiffi entity."""
+        self._id = generateUniqueId(device, metric)
+        self._device_info = {
+            "connections": {
+                (device_registry.CONNECTION_NETWORK_MAC, device.mac_address)
+            },
+            "identifiers": {(DOMAIN, device.mac_address)},
+            "manufacturer": "stall.biz",
+            "name": f"{device.moduletype} {device.mac_address}",
+            "model": device.moduletype,
+            "sw_version": device.sw_version,
+        }
+        self._name = metric.description
+        self._expiration_date = None
+        self._value = None
+
+    async def async_added_to_hass(self):
+        """Entity has been added to hass."""
+        async_dispatcher_connect(
+            self.hass, UPDATE_ENTITY_SIGNAL + self._id, self._update_value_callback
+        )
+        async_dispatcher_connect(
+            self.hass, CHECK_ENTITIES_SIGNAL, self._check_expiration_date
+        )
+
+    @property
+    def should_poll(self):
+        """Disable polling because data driven ."""
+        return False
+
+    @property
+    def device_info(self):
+        """Return wiffi device info which is shared between all entities of a device."""
+        return self._device_info
+
+    @property
+    def unique_id(self):
+        """Return unique id for entity."""
+        return self._id
+
+    @property
+    def name(self):
+        """Return entity name."""
+        return self._name
+
+    @property
+    def available(self):
+        """Return true if value is valid."""
+        return self._value is not None
+
+    def reset_expiration_date(self):
+        """Reset value expiration date.
+
+        Will be called by derived classes after a value update has been received.
+        """
+        self._expiration_date = datetime.now() + timedelta(minutes=3)
```

Comment:
Please use utctime unless we need to use local time to interface with an api. And please use our dt util as much as possible. We have `utcnow` available there.